include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1107,302 @@ AlignedVector<T>::fill(const T &value)
+  // documentation, and that we do not actually need. So we "abuse" it a bit: On
+  // the shmem root, we put the array size into it. Later on, the remaining
+  // processes can query the shmem root's value of 'disp_unit', and so will be
+  // able to learn about the array size that way.
+  MPI_Win        shmem_window;
+  T *            base_ptr;
+  const MPI_Aint align_by = 64;
+
+  {
+    const MPI_Aint alloc_size =
+      (is_shmem_root ?
+         ((size() * sizeof(T) + align_by - 1) / sizeof(T)) * sizeof(T) :
+         0);
+    const int disp_unit = (is_shmem_root ? size() : 1);
+
+    const int ierr = MPI_Win_allocate_shared(alloc_size,
+                                             disp_unit,
+                                             MPI_INFO_NULL,
+                                             shmem_group_communicator,
+                                             &base_ptr,
+                                             &shmem_window);
+    AssertThrowMPI(ierr);
+  }
+
+
+  // **** Step 4 ****
+  // The next step is to teach all non-shmem root processes what the pointer to
+  // the array is that the shmem-root created. MPI has a nifty way for this
+  // given that only a single process actually allocated memory in the window:
+  // When calling MPI_Win_shared_query, the MPI documentation says that
+  // "When rank is MPI_PROC_NULL, the pointer, disp_unit, and size returned are
+  // the pointer, disp_unit, and size of the memory segment belonging the lowest
+  // rank that specified size > 0. If all processes in the group attached to the
+  // window specified size = 0, then the call returns size = 0 and a baseptr as
+  // if MPI_ALLOC_MEM was called with size = 0."
+  //
+  // This will allow us to obtain the pointer to the shmem root's memory area,
+  // which is the only we one we care about. (None of the other processes have
+  // even allocated any memory.) But this will also retrieve the shmem root's
+  // disp_unit, which in step 3 above we have abused to pass along the number of
+  // elements in the array.
+  //
+  // We don't need to do this on the shmem root process: This process has
+  // already gotten its base_ptr correctly set above, and we can determine the
+  // array size by just calling size().
+  unsigned int array_size =
+    (is_shmem_root ? size() : numbers::invalid_unsigned_int);
+  if (is_shmem_root == false)
+    {
+      int       disp_unit;
+      MPI_Aint  alloc_size; // not actually used
+      const int ierr = MPI_Win_shared_query(
+        shmem_window, MPI_PROC_NULL, &alloc_size, &disp_unit, &base_ptr);
+      AssertThrowMPI(ierr);
+
+      array_size = disp_unit;
+    }
+
+
+  // **** Step 5 ****
+  // Now that all processes know the address of the space that is visible to
+  // everyone, we need to figure out whether it is properly aligned and if not,
+  // find the next aligned address
+
+  // TODO: Align the base_ptr via std::align
```

Comment:
is this still missing?

Reply:
Yes. I just don't quite know how to do it. I have to admit that I don't quite understand the syntax of `std::align.`